File: alg3/alg3-tree.py
class Node:
    def __init__(self, index = 0, depth = 0):
        self.index = index;
        self.depth = depth;
        self.parent = None;
        
class Tree:
    def __init__(self, size):
        self.size = size
        self.max_depth = 0
        self.deepest_leaf = None
        self.nodes = [Node() for i in range(size)]
        self.root = self.nodes[0];
        

filename = "alg3-data.txt"
f = open(filename)

n = int(next(f))
print("n =", n)

tr = Tree(n + 1)

for i in range(1, n):
    nd = tr.nodes[i]
    nd.index = i

    pi = int(next(f))
    pnd = tr.nodes[pi]
    nd.parent = pnd
    nd.depth = pnd.depth + 1

    if nd.depth > tr.max_depth:
        tr.max_depth = nd.depth;
        tr.deepest_leaf = nd;

print(tr.deepest_leaf.index)

File: alg3/alg3-tree.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <algorithm>

using namespace std;

struct Node {
    int index; // Индекс узла
    int depth;
    vector<Node*> children; // Вектор детей узла
    Node *parent;
    
};

ostream& operator<<(ostream& os, const Node* node) {
    os << "Глубина узла " << node->depth << " индекса " << node->index << endl;
    for (const Node* child : node->children) {
        os << "   Дочерний узел: " << child->depth << " индекса " << child->index << endl;
    }
    return os;
}

int main() {
    string filename = "alg3-data.txt";
    ifstream fin{filename};

    int n = 0;
    fin >> n;
    cout << "n = " << n << endl;

    Node *tree = new Node[n + 1] {};
    Node *root = tree; 

    int max_depth = 0;
    Node *deepest_leaf = nullptr;

    for(int i = 1; i <= n; i++) {
        Node *nd = &tree[i];

        nd->index = i;

        int pi;
        fin >> pi;
        Node *pnd = &tree[pi];

        nd->parent = pnd;
        pnd->children.push_back(nd);
        nd->depth = pnd->depth + 1;

        if(nd->depth > max_depth) {
            max_depth = nd->depth;
            deepest_leaf = nd;
        }
    }
    delete[] tree;
    cout << deepest_leaf->index;
}
